test/Conversion/LinalgToXsmm/linalg-to-binary.mlir

```diff
@@ -377,3 +377,119 @@ func.func @sub_bcast_row_1(%arg0: memref<256x1024xf32>, %arg1: memref<256xf32>)
+        linalg.yield %19 : f32
+  }
+  return 
+}
+
+// CHECK-LABEL: mul_bcast_scalar
+// CHECK-SAME: %[[ARG0:.+]]: memref<256x1024xf32>, %[[ARG1:.+]]: f32, %[[ARG2:.+]]: memref<256x1024xf32>
+// CHECK: %[[DIS:.+]] = xsmm.binary.dispatch mul [256, 1024, 1024, 1, 1024] flags = (bcast_scalar_in1) data_type = f32
+// CHECK: xsmm.binary mul(data_type = f32, %[[DIS]], %[[ARG0]], %[[ARG1]], %[[ARG2]])
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+#map1 = affine_map<(d0, d1) -> (d1)>
+
+func.func @mul_bcast_col(%arg0: memref<256x1024xf32>, %arg1: memref<1024xf32>, %arg2: memref<256x1024xf32>) {
+  linalg.generic {
+    indexing_maps = [#map, #map1, #map], 
+    iterator_types = ["parallel", "parallel"]}
+    ins(%arg0, %arg1 : memref<256x1024xf32>, memref<1024xf32>)
+    outs(%arg2: memref<256x1024xf32>) {
+      ^bb0(%in: f32, %in_4: f32, %out: f32):
+        %19 = arith.mulf %in, %in_4 : f32
+        linalg.yield %19 : f32
+  }
+  return 
+}
+
+// CHECK-LABEL: mul_bcast_col
+// CHECK-SAME: %[[ARG0:.+]]: memref<256x1024xf32>, %[[ARG1:.+]]: memref<1024xf32>, %[[ARG2:.+]]: memref<256x1024xf32>
+// CHECK: %[[DIS:.+]] = xsmm.binary.dispatch mul [256, 1024, 1024, 1024, 1024] flags = (bcast_col_in1) data_type = f32
+// CHECK: xsmm.binary mul(data_type = f32, %[[DIS]], %[[ARG0]], %[[ARG1]], %[[ARG2]])
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+#map1 = affine_map<(d0, d1) -> (d0, 0)>
+
+func.func @mul_bcast_row(%arg0: memref<256x1024xf32>, %arg1: memref<256x1xf32>, %arg2: memref<256x1024xf32>) {
+  linalg.generic {
+    indexing_maps = [#map, #map1, #map], 
+    iterator_types = ["parallel", "parallel"]}
+    ins(%arg0, %arg1 : memref<256x1024xf32>, memref<256x1xf32>)
+    outs(%arg2: memref<256x1024xf32>) {
+      ^bb0(%in: f32, %in_4: f32, %out: f32):
+        %19 = arith.mulf %in, %in_4 : f32
+        linalg.yield %19 : f32
+  }
+  return 
+}
+
+// CHECK-LABEL: mul_bcast_row
+// CHECK-SAME: %[[ARG0:.+]]: memref<256x1024xf32>, %[[ARG1:.+]]: memref<256x1xf32>, %[[ARG2:.+]]: memref<256x1024xf32>
+// CHECK: %[[DIS:.+]] = xsmm.binary.dispatch mul [256, 1024, 1024, 1, 1024] flags = (bcast_row_in1) data_type = f32
+// CHECK: xsmm.binary mul(data_type = f32, %[[DIS]], %[[ARG0]], %[[ARG1]], %[[ARG2]])
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+#map1 = affine_map<(d0, d1) -> (d0)>
+
+func.func @mul_bcast_row_1(%arg0: memref<256x1024xf32>, %arg1: memref<256xf32>) {
+  linalg.generic {
+    indexing_maps = [#map, #map1, #map],
+    iterator_types = ["parallel", "parallel"]}
+    ins(%arg0, %arg1 : memref<256x1024xf32>, memref<256xf32>)
+    outs(%arg0 : memref<256x1024xf32>) {
+    ^bb0(%in: f32, %in_6: f32, %out: f32):
+      %6 = arith.mulf %in, %in_6 : f32
+      linalg.yield %6 : f32
+  }
+  return
+}
+
+// CHECK-LABEL: mul_bcast_row_1
+// CHECK-SAME: %[[ARG0:.+]]: memref<256x1024xf32>, %[[ARG1:.+]]: memref<256xf32>
+// CHECK: linalg.generic
+// CHECK-NOT: xsmm.binary sub
+// We need to lift this restriction:
```

Comment:
This looks like a bad broadcast to me, not something we want to cover. Both `<1024>` and `<256x1>` are covered above.

Reply:
Why? It is perfectly defined by the affine map. If I am not mistaken we can end up in this situation if we fold unit dims in linalg.

Reply:
Ah, good point! Ignore me.